File: leetcode/393.utf-8-validation.py
class Solution(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """
        if data is None or len(data) == 0:
            return False
        
        i = 0
        while i < len(data):
            n = self.byteLength(data[i])
            if n == 0 or n + i - 1 >= len(data):
                return False
            
            for j in range(i + 1, n + i):
                if (data[j] >> 6) != 0b10:
                    return False
            
            i = i + n
            
        return True
        
    def byteLength(self, n):
        if (n >> 7) == 0b0:
            return 1
        if (n >> 5) == 0b110:
            return 2
        if (n >> 4) == 0b1110:
            return 3
        if (n >> 3) == 0b11110:
            return 4
        return 0

File: leetcode/393.utf-8-validation.cpp
class Solution {
private:
    int byteLength(int n)
    {
        if ((n >> 7) == 0b0)
        {
            return 1;
        }
        else if ((n >> 5) == 0b110)
        {
            return 2;
        }
        else if ((n >> 4) == 0b1110)
        {
            return 3;
        }
        else if ((n >> 3) == 0b11110)
        {
            return 4;
        }
        
        return 0;
    }
public:
    bool validUtf8(vector<int>& data) {
        
        if (data.size() == 0)
        {
            return false;
        }
        
        int i = 0;
        
        while (i < data.size())
        {
            int n = data[i];
            int len = byteLength(n);
            
            if (len == 0 || i + len - 1 >= data.size())
            {
                return false;
            }
            
            for (auto j = i + 1; j <= i + len - 1; ++j)
            {
                if (data[j] >> 6 != 0b10)
                {
                    return false;
                }
            }
            
            i = i + len;
        }
        
        return true;
    }
};